public/app/features/explore/ShortLinkButtonMenu.tsx

```diff
@@ -0,0 +1,151 @@
+import React, { useState } from 'react';
+
+import { IconName } from '@grafana/data';
+import { reportInteraction, config } from '@grafana/runtime';
+import { ToolbarButton, Dropdown, Menu, Stack, ToolbarButtonRow, MenuGroup } from '@grafana/ui';
+import { t } from 'app/core/internationalization';
+import { copyStringToClipboard } from 'app/core/utils/explore';
+import { createAndCopyShortLink } from 'app/core/utils/shortLinks';
+import { useSelector } from 'app/types';
+
+import { selectPanes } from './state/selectors';
+import { constructAbsoluteUrl } from './utils/links';
+
+interface ShortLinkGroupData {
+  key: string;
+  label: string;
+  items: ShortLinkMenuItemData[];
+}
+
+interface ShortLinkMenuItemData {
+  key: string;
+  label: string;
+  icon: IconName;
+  getUrl: Function;
+  shorten: boolean;
+}
+
+const defaultMode: ShortLinkMenuItemData = {
+  key: 'copy-link',
+  label: t('explore.toolbar.copy-shortened-link', 'Copy shortened URL'),
+  icon: 'share-alt',
+  getUrl: () => undefined,
+  shorten: true,
+};
+
+export function ShortLinkButtonMenu() {
+  const panes = useSelector(selectPanes);
+  const [isOpen, setIsOpen] = useState(false);
+  const [lastSelected, setLastSelected] = useState(defaultMode);
+  const onCopyLink = (shorten: boolean, url?: string) => {
+    if (shorten) {
+      createAndCopyShortLink(url || global.location.href);
+      reportInteraction('grafana_explore_shortened_link_clicked');
+    } else {
+      copyStringToClipboard(
```

Comment:
It'd be nice if generating URLs was in a util outside of this component.

Reply:
The difficult process of generating an absolute URL is abstracted out of this component and I don't wish to abstract this any more than it is already. We can follow up with this if we want.